File: app/src/main/java/com/mohsenoid/rickandmorty/view/base/BaseActivity.java
package com.mohsenoid.rickandmorty.view.base;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.mohsenoid.rickandmorty.RickAndMortyApplication;
import com.mohsenoid.rickandmorty.injection.DependenciesProvider;

public abstract class BaseActivity extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        DependenciesProvider dependenciesProvider = ((RickAndMortyApplication) getApplicationContext()).getDependenciesProvider();

        injectDependencies(savedInstanceState, dependenciesProvider);
        super.onCreate(savedInstanceState);
    }

    protected abstract void injectDependencies(@Nullable Bundle savedInstanceState, DependenciesProvider dependenciesProvider);
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/view/base/BaseActivity.kt
package com.mohsenoid.rickandmorty.view.base

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.mohsenoid.rickandmorty.util.dispatcher.DispatcherProvider
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import org.koin.android.ext.android.inject
import kotlin.coroutines.CoroutineContext

abstract class BaseActivity : AppCompatActivity(), CoroutineScope {

    private val dispatcherProvider: DispatcherProvider by inject()

    private lateinit var job: Job
    override lateinit var coroutineContext: CoroutineContext

    override fun onCreate(savedInstanceState: Bundle?) {
        job = SupervisorJob()
        coroutineContext = dispatcherProvider.mainDispatcher + job

        super.onCreate(savedInstanceState)
    }

    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }
}
